baselines/fedexp/fedexp/main.py

```diff
@@ -0,0 +1,110 @@
+"""Main script for running FedExp."""
+
+import flwr as fl
+import hydra
+import numpy as np
+from hydra.core.hydra_config import HydraConfig
+from hydra.utils import instantiate
+from omegaconf import DictConfig, OmegaConf
+
+from fedexp import client, server, utils
+from fedexp.dataset import load_datasets
+from fedexp.utils import get_parameters, seed_everything
+
+
+@hydra.main(config_path="conf", config_name="base", version_base=None)
+def main(cfg: DictConfig) -> None:
+    """Run the baseline.
+
+    Parameters
+    ----------
+    cfg : DictConfig
+        An omegaconf object that stores the hydra config.
+    """
+    print(OmegaConf.to_yaml(cfg))
+
+    seed_everything(cfg.seed)
+
+    train_loaders, test_loader = load_datasets(
+        config=cfg.dataset_config,
+        num_clients=cfg.num_clients,
+        batch_size=cfg.batch_size,
+        partition_equal=True,
+    )
+
+    data_ratios = np.zeros(cfg.num_clients)
+    for i in range(cfg.num_clients):
+        data_ratios[i] = len(train_loaders[i])
+    data_ratios /= np.sum(data_ratios)
+
+    client_fn = client.gen_client_fn(
+        train_loaders=train_loaders,
+        model=cfg.model,
+        num_epochs=cfg.num_epochs,
+        args={"data_ratio": data_ratios, "device": cfg.client_device},
+    )
+
+    evaluate_fn = server.gen_evaluate_fn(
+        test_loader=test_loader, model=cfg.model, device=cfg.server_device
+    )
+
+    def get_on_fit_config():
+        def fit_config_fn(server_round: int):
+            fit_config = OmegaConf.to_container(cfg.hyperparams, resolve=True)
+            fit_config["curr_round"] = server_round
+            cfg.hyperparams.eta_l *= cfg.hyperparams.decay
+            return fit_config
+
+        return fit_config_fn
+
+    net_glob = instantiate(cfg.model)
+
+    strategy = instantiate(
+        cfg.strategy,
+        evaluate_fn=evaluate_fn,
+        on_fit_config_fn=get_on_fit_config(),
+        initial_parameters=fl.common.ndarrays_to_parameters(get_parameters(net_glob)),
+        net_glob=net_glob,
+        epsilon=cfg.hyperparams.epsilon,
+        decay=cfg.hyperparams.decay,
+    )
+
+    history = fl.simulation.start_simulation(
+        client_fn=client_fn,
+        num_clients=cfg.num_clients,
+        config=fl.server.ServerConfig(num_rounds=cfg.num_rounds),
+        strategy=strategy,
+        client_resources={
+            "num_cpus": cfg.client_resources.num_cpus,
+            "num_gpus": cfg.client_resources.num_gpus,
+        },
+        ray_init_args={
+            "ignore_reinit_error": True,
+            "include_dashboard": False,
+            "num_cpus": 12,
+            "num_gpus": 1,
+            "_memory": 100*1024*1024*1024,
```

Comment:
Are all these necessary?

Reply:
it was necessary for our server to detect the resources.

Reply:
Ok. I'd say better not to leave this hardcoded  (i.e. `num_cpus`, `num_gpus`, `_memory`)because for the final version of your code because this will make the `simulation_start` fail if someone wants to run your `FedExp` but doesn't have access to that many resources. Similarly, if someone has more, their system can potentially be underutilised. 

Happy to leave this there until before merging with `main`.